### PR TITLE
Core.py removing 'type: ignore' hints

### PR DESCRIPTION
**Pull request by michael-gendy-mention-me**
As part of [this issue](https://github.com/dmlc/xgboost/issues/6496#issuecomment-1558022930), we want to remove some of the `type: ignore` hints to make the code more rigorous. This PR will aim to remove all the `type: ignore` hints in core.py

### CODE REVIEW

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -82,13 +82,11 @@ def from_pystr_to_cstr(data: Union[str, List[str]]) -> Union[bytes, ctypes.Array
...
     if isinstance(data, list):
-        pointers: ctypes.Array[ctypes.c_char_p] = (ctypes.c_char_p * len(data))()
-        data_as_bytes = [bytes(d, "utf-8") for d in data]
-        pointers[:] = data_as_bytes  # type: ignore
+        data_as_bytes: List[bytes] = [bytes(d, "utf-8") for d in data]
+        pointers: ctypes.Array[ctypes.c_char_p] = (ctypes.c_char_p * len(data_as_bytes))(*data_as_bytes)
         return pointers
     raise TypeError()
 
-
```
Comment (michael-gendy-mention-me) [inline python-package/xgboost/core.py:91]: This is out of my wheelhouse somewhat, but to me it looks like we're initialising an empty ctypes.Array (`pointers`), then adding `data_as_bytes` to it which was creating a mismatch between the List[bytes] and ctypes.Array[ctypes.c_char_p]. I've just added the `data_as_bytes` into `pointers` at initialisation, does this make sense?

**Comment by michael-gendy-mention-me** [comment]
Just committed the linting changes from https://github.com/dmlc/xgboost/actions/runs/5069030065/jobs/9121890991

File: python-package/xgboost/core.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/core.py]: To be honest, I don't quite understand it either. Need to look it up. ;-)

**Review by trivialfis — APPROVED** [review]
Looks good to me, thank you for removing the workarounds!

**Comment by michael-gendy-mention-me** [comment]
Apologies I've been away for a number of days. Thanks a lot for the review, merge and help finding something to work on @trivialfis 😄

**Pull request merged**